extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java

```diff
@@ -18,28 +22,66 @@
+     * annotations on a MicroProfile REST Client method. In that case, we have no other choice but to rely on reflection (with
+     * underlying synchronized blocks which are bad for performances) to retrieve the interceptor bindings.
+     */
+    protected <T extends Annotation> CacheInterceptionContext<T> getInterceptionContext(InvocationContext invocationContext,
+            Class<T> interceptorBindingClass) {
+        return getArcCacheInterceptionContext(invocationContext, interceptorBindingClass)
+                .orElse(getNonArcCacheInterceptionContext(invocationContext, interceptorBindingClass));
+    }
+
+    private <T extends Annotation> Optional<CacheInterceptionContext<T>> getArcCacheInterceptionContext(
+            InvocationContext invocationContext, Class<T> interceptorBindingClass) {
+        Set<Annotation> bindings = InterceptorBindings.getInterceptorBindings(invocationContext);
+        if (bindings == null) {
+            // This should only happen when the interception is not managed by Arc.
+            return Optional.empty();
+        }
+        CacheInterceptionContext<T> result = new CacheInterceptionContext<>();
+        for (Annotation binding : bindings) {
+            if (binding instanceof CacheKeyParameterPositions) {
+                short[] positions = ((CacheKeyParameterPositions) binding).value();
+                result.setCacheKeyParameterPositions(positions);
+            } else if (interceptorBindingClass.isInstance(binding)) {
+                result.getInterceptorBindings().add(cast(binding, interceptorBindingClass));
             }
         }
-        return bindings;
+        return Optional.of(result);
     }
 
-    protected <T extends Annotation> T getInterceptorBinding(InvocationContext context, Class<T> bindingClass) {
-        return getInterceptorBindings(context, bindingClass).get(0);
+    private <T extends Annotation> CacheInterceptionContext<T> getNonArcCacheInterceptionContext(
+            InvocationContext invocationContext, Class<T> interceptorBindingClass) {
+        CacheInterceptionContext<T> result = new CacheInterceptionContext<>();
+        for (Annotation annotation : invocationContext.getMethod().getAnnotations()) {
+            if (interceptorBindingClass.isInstance(annotation)) {
+                result.getInterceptorBindings().add(cast(annotation, interceptorBindingClass));
+            }
+        }
+        Parameter[] parameters = invocationContext.getMethod().getParameters();
+        if (parameters.length > 0) {
+            List<Short> positions = new ArrayList<>();
+            for (short i = 0; i < parameters.length; i++) {
+                if (parameters[i].isAnnotationPresent(CacheKey.class)) {
+                    positions.add(i);
+                }
+            }
+            short[] positionsArray = new short[positions.size()];
+            for (int i = 0; i < positions.size(); i++) {
+                positionsArray[i] = positions.get(i);
+            }
```

Comment:
I know this is not pretty, but there's a reason behind that: avoid lambdas in runtime code.